dot/core/service.go

```diff
@@ -454,258 +382,37 @@ func (s *Service) handleReceivedMessage(msg network.Message) (err error) {
-	subchain, err := s.blockState.SubChain(startHash, endHash)
-	if err != nil {
-		return nil, err
-	}
-
-	if len(subchain) > int(maxResponseSize) {
-		subchain = subchain[:maxResponseSize]
-	}
-
-	responseData := []*types.BlockData{}
-
-	for _, hash := range subchain {
-
-		blockData := new(types.BlockData)
-		blockData.Hash = hash
-
-		// set defaults
-		blockData.Header = optional.NewHeader(false, nil)
-		blockData.Body = optional.NewBody(false, nil)
-		blockData.Receipt = optional.NewBytes(false, nil)
-		blockData.MessageQueue = optional.NewBytes(false, nil)
-		blockData.Justification = optional.NewBytes(false, nil)
-
-		// header
-		if (blockRequest.RequestedData & 1) == 1 {
-			retData, err := s.blockState.GetHeader(hash)
-			if err == nil && retData != nil {
-				blockData.Header = retData.AsOptional()
-			}
-		}
-		// body
-		if (blockRequest.RequestedData&2)>>1 == 1 {
-			retData, err := s.blockState.GetBlockBody(hash)
-			if err == nil && retData != nil {
-				blockData.Body = retData.AsOptional()
-			}
-		}
-		// receipt
-		if (blockRequest.RequestedData&4)>>2 == 1 {
-			retData, err := s.blockState.GetReceipt(hash)
-			if err == nil && retData != nil {
-				blockData.Receipt = optional.NewBytes(true, retData)
-			}
-		}
-		// message queue
-		if (blockRequest.RequestedData&8)>>3 == 1 {
-			retData, err := s.blockState.GetMessageQueue(hash)
-			if err == nil && retData != nil {
-				blockData.MessageQueue = optional.NewBytes(true, retData)
-			}
-		}
-		// justification
-		if (blockRequest.RequestedData&16)>>4 == 1 {
-			retData, err := s.blockState.GetJustification(hash)
-			if err == nil && retData != nil {
-				blockData.Justification = optional.NewBytes(true, retData)
-			}
-		}
-
-		responseData = append(responseData, blockData)
-	}
-
-	return &network.BlockResponseMessage{
-		ID:        blockRequest.ID,
-		BlockData: responseData,
-	}, nil
-}
-
-// ProcessBlockResponseMessage attempts to validate and add the block to the
-// chain by calling `core_execute_block`. Valid blocks are stored in the block
-// database to become part of the canonical chain.
-func (s *Service) ProcessBlockResponseMessage(msg network.Message) error {
-	log.Debug("[core] received BlockResponseMessage")
-	s.respOut <- msg.(*network.BlockResponseMessage)
-
-	return s.checkForRuntimeChanges()
-}
-
-// checkForRuntimeChanges checks if changes to the runtime code have occurred; if so, load the new runtime
-func (s *Service) checkForRuntimeChanges() error {
-	currentCodeHash, err := s.storageState.LoadCodeHash()
-	if err != nil {
-		return err
-	}
-
-	if !bytes.Equal(currentCodeHash[:], s.codeHash[:]) {
-		code, err := s.storageState.LoadCode()
-		if err != nil {
-			return err
-		}
-
-		s.rt.Stop()
+func (s *Service) handleBabeSession() {
+	for {
+		// wait for BABE epoch to complete
+		<-s.epochDone
 
-		s.rt, err = runtime.NewRuntime(code, s.storageState, s.keys)
+		// verify first block was set during current epoch
+		err := s.verifyFirstBlock()
```

Comment:
this is what is throwing the error in the logs and needs #662 in order to be resolved